File: Kotlin/BlackJack/src/unitTests/DealerTest.kt
package unitTests

import BlackJack.entity.Dealer
import BlackJack.observer.DealerObserver
import org.junit.Test

import org.junit.Assert.*

class DealerTest {
    @Test
    fun giveCardAndNoticeLowDeckTest() {

        val testDealerObserver = TestDealerObserver()

        val dealer = Dealer(numberOfDecksUsed = 1, observer = testDealerObserver)

        testDealerObserver.dealer = dealer

        assertEquals(52, dealer.getDeckPoolSize())

        dealer.giveCard()

        assertEquals(51, dealer.getDeckPoolSize())

        (1..25).forEach { _ ->
            dealer.giveCard()
        }

        assertEquals(26, dealer.getDeckPoolSize())

        dealer.giveCard()

        assertEquals(25, dealer.getDeckPoolSize())

        dealer.giveCard()

        assertEquals(51, dealer.getDeckPoolSize())

    }
}

class TestDealerObserver: DealerObserver{

    var dealer: Dealer? = null

    override fun noticeLowDeck() {

        dealer?.createNewDeck()

    }

}

File: Lua/BlackJack/blackjack/test/dealerTest.lua


local Dealer = require "entity.Dealer"

local testObserver = {
    dealer = {}
}

function testObserver:noticeLowDeck()
    self.dealer:createNewDeck()
end

function testGiveCardAndNoticeLowDeck()
    
    local dealer = Dealer(1, testObserver)

    testObserver.dealer = dealer
    
    lu.assertEquals(#dealer.deckPool, 52)
    
    dealer:giveCard()
    
    lu.assertEquals(#dealer.deckPool, 51)

    for i = 1, 25, 1 do
        dealer:giveCard()
    end
    
    lu.assertEquals(#dealer.deckPool, 26)
    
    dealer:giveCard()
    
    lu.assertEquals(#dealer.deckPool, 25)
    
    dealer:giveCard()
    
    lu.assertEquals(#dealer.deckPool, 51)

end